### PR TITLE
crypto: complete #375 audit — checked-in matrix, checklist, constant-time helper routing

### PR DESCRIPTION
**Pull request by Jaruso**
## Summary

This closes out #375 (the remaining side-channel/secret-lifetime/observability audit work, after #549 already landed the secret-lifecycle-cleanup portion). #490 (the provider-ownership architecture #375 was gated on) is closed, so this audit is scoped against its final architecture.

- Adds `docs/CRYPTO_SECURITY_AUDIT.md` — the required checked-in audit matrix (module/function, value, classification, operation, attacker-observability, owner/lifetime, cleanup paths, required treatment, status, rationale), covering TLS handshake secrets, QUIC, ticket protection/resumption, a dedicated resumption/ticket-oracle review section, a table of representative public comparisons intentionally left non-constant-time, observability, the provider/ownership boundary, toolchain/platform assumptions, and deferred findings.
- Adds `docs/SECURITY_REVIEW_CHECKLIST.md` — the reusable review checklist for future security-sensitive changes (#375's 8-question list, expanded).
- Migrates the raw `std.crypto.timing_safe`/`std.crypto.secureZero` call sites this audit found over secret/authentication-derived material to the canonical `crypto.secrets`/`crypto.provider` helpers:
  - TLS PSK binder verification (`src/tls/pre_shared_key.zig`)
  - TLS Finished verification, client and server (`src/tls/tls13_backend.zig`)
  - QUIC Retry integrity tag — both the live production check (`src/quic/packet.zig`, called from `connection.zig`) and an unreachable KAT-fixture-only duplicate the audit surfaced (`src/quic/path.zig`)
  - QUIC PATH_RESPONSE validation (`src/quic/path.zig`)
  - RSA-PSS final hash comparison (`src/crypto/rsa.zig`)
  - Zeroization helper-routing consistency in `src/tls/ticket_key_snapshot.zig` and `src/tls/sni_provider.zig`

No comparison or zeroization *semantics* change anywhere — every touched site already used a constant-time compare or a functionally equivalent zero call; this is canonical-helper routing plus the audit trail proving it's correct, not a behavior fix.

Every other location the audit reviewed — ticket-protection key/fingerprint equality, the QUIC Retry-token key ring and stateless-reset key lifecycle, session cache bearer-handle comparison, credentials/identity-loader secret cleanup — was already compliant, largely from #549's earlier secret-lifecycle hardening (re-verified here, not re-fixed).

## Deferred findings

Per #375's acceptance criteria, both are tracked before this closes the issue:
- #554 — extend `scripts/audit_crypto_boundary.zig` to guard against regression of the findings this PR fixes.
- #555 — consolidate the duplicate Retry-integrity-tag implementation in `packet.zig`/`path.zig`.

## Test plan
- [x] `zig build test` — 2977 tests, 2976 passed / 1 pre-existing skip (unrelated to this change)
- [x] `zig fmt --check build.zig src/ tests/` — clean
- [x] `zig build audit-crypto-boundary` (part of `zig build test`) — passes; no new direct-crypto call sites introduced
- [x] Existing binder/Finished/Retry/PATH_RESPONSE/RSA-PSS accept-and-reject test coverage exercises every migrated call site (see "Regression tests" section of the audit doc for the specific test names)

Closes #375

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by Jaruso — COMMENTED** [review]
Blocking review finding: the helper-routing changes themselves look behavior-preserving, and the audit/checklist cover most of #375's required surfaces well. However, the zeroization audit is internally inconsistent with the implementation it approves.

`docs/CRYPTO_SECURITY_AUDIT.md` explicitly states that `secureZero(...); allocator.free(...)` is not an acceptable heap-secret cleanup path and that `secureZeroAndFree`/raw-free semantics are required so the backing allocator observes zeroed bytes. This PR then changes several `ticket_key_snapshot.zig` sites only from `std.crypto.secureZero` to `provider.secureZero` while retaining ordinary `Allocator.free`/`ArrayList.deinit`, and the canonical `BoundedSecret.deinit` still has the same pattern. That means the acceptance criteria "ad hoc zeroization is removed" and "secret owners ... implement cleanup" are not actually satisfied under the audit's own stated model.

Please make the implementation and audit agree before closing #375. At minimum, fix every secret-bearing heap cleanup identified by this audit (including `OwnedSnapshot.key_storage`, file buffers, serialized snapshot buffers, parse-error storage, and `BoundedSecret`) using an alignment-correct zero-and-raw-free path, and add allocator-observable regression coverage. For typed allocations, do not pass a byte-cast slice to the existing `secureZeroAndFree`, because it hardcodes `alignOf(u8)`; either add a generic aligned helper or perform the equivalent `secureZero(std.mem.sliceAsBytes(...))` followed by `allocator.rawFree(..., .fromByteUnits(@alignOf(T)), ...)` in one reviewed abstraction. Then update the matrix to explicitly list these owners and their cleanup disposition.

I am leaving this as a comment rather than a formal `REQUEST_CHANGES` review because GitHub does not allow requesting changes on your own PR.

File: src/tls/ticket_key_snapshot.zig [hunk 1]
```diff
@@ -79,7 +79,7 @@ pub const OwnedSnapshot = struct {
     key_storage: []KeyStorage,
 
     pub fn deinit(self: *OwnedSnapshot) void {
-        for (self.key_storage) |*storage| std.crypto.secureZero(u8, &storage.bytes);
+        for (self.key_storage) |*storage| provider.secureZero(&storage.bytes);
```
Comment (Jaruso) [inline src/tls/ticket_key_snapshot.zig]: This still uses the cleanup pattern the new audit explicitly declares insufficient: the next line calls ordinary `Allocator.free`, whose wrapper may overwrite/poison the zeroed bytes before the backing allocator sees them. The same issue appears in `loadFromFile`, `reserveNonceLeasesInFile`, and the parse `errdefer`, and `crypto.secrets.BoundedSecret.deinit` remains unchanged with this pattern.

Please route secret-bearing heap releases through an alignment-correct zero-and-raw-free abstraction. For example, a generic helper could zero `std.mem.sliceAsBytes(buffer)` and then call `allocator.rawFree(..., .fromByteUnits(@alignOf(T)), @returnAddress())`; the existing byte-only `secureZeroAndFree` cannot safely free `[]KeyStorage` because it hardcodes byte alignment. Add a backing-allocator test that verifies the bytes observed at free time are zero, and record these owners explicitly in the audit matrix.

**Review by Jaruso — COMMENTED** [review]
Re-reviewed the updated head against #375 and the previous blocking finding.

The cleanup inconsistency is resolved correctly:

- `secureZeroAndFreeAligned` wipes the byte view and forwards the element alignment to `rawFree`.
- `BoundedSecret.deinit` now uses the raw-free helper instead of zeroing and then calling ordinary `Allocator.free`.
- `OwnedSnapshot.deinit` and the parse error path use the aligned helper for `[]KeyStorage`.
- Raw file bytes and the serialization buffer use `secureZeroAndFree`; the serialization path correctly wipes/frees `allocatedSlice()` rather than only `items`, so unused capacity is covered and the slice passed to `rawFree` matches the allocation.
- The audit matrix now accurately describes the behavior change and affected owners.
- The new allocator-observable regression tests cover the shared helper, `BoundedSecret`, and the real snapshot parse/deinit path.

I also rechecked the original constant-time helper migrations and the issue acceptance criteria. I found no additional blocking defects. CI is green on commit `e34928e`.

The previous review thread is resolved. This is ready to merge from my review perspective. GitHub does not permit an approval review on a PR owned by this account, so this is submitted as a comment review.
